src/main/java/com/aws/greengrass/clientdevices/auth/ClientDevicesAuthService.java

```diff
@@ -200,19 +215,51 @@ protected void install() throws InterruptedException {
 
             if (whatHappened == WhatHappened.initialized || node == null) {
                 updateDeviceGroups(whatHappened, deviceGroupTopics);
-                updateCAType(caTypeTopic);
+                updateCertificateAuthority();
             } else if (node.childOf(DEVICE_GROUPS_TOPICS)) {
                 updateDeviceGroups(whatHappened, deviceGroupTopics);
-            } else if (node.childOf(CA_TYPE_TOPIC)) {
-                if (caTypeTopic.getOnce() == null) {
+            } else if (node.childOf(CERTIFICATE_AUTHORITY_TOPIC)) {
+                if (certificateAuthorityTopic.getOnce() == null) {
                     return;
                 }
-                updateCAType(caTypeTopic);
+                updateCertificateAuthority();
             }
         });
 
     }
 
+    private void updateCertificateAuthority() {
+        String caKeyPath = Coerce.toString(config.lookup(CERTIFICATE_AUTHORITY_TOPIC, CA_PRIVATE_KEY_PATH).dflt(""));
+        String caCertPath =
+                Coerce.toString(config.lookup(CERTIFICATE_AUTHORITY_TOPIC, CA_CERTIFICATE_FILE_PATH).dflt(""));
+        Topic caType = config.lookup(CERTIFICATE_AUTHORITY_TOPIC, CA_TYPE_TOPIC);
+
+        // If the key and cert path of CA are empty, generate a CA
+        if (Utils.isEmpty(caKeyPath) || Utils.isEmpty(caCertPath)) {
+            updateCAType(caType);
+        } else {
+            // Get CA from URI and upload
+            URI privateKeyUri = SecurityService.uriFromPossibleFileURIString(caKeyPath);
+            URI certificateUri = SecurityService.uriFromPossibleFileURIString(caCertPath);
```

Comment:
since these are new, we can be strict here. We can ensure that these are absolutely URIs and not just file paths. We couldn't do that in the Nucleus for compatibility, but that's not a concern here. Recommend trying to parse as a URI and then handling failure appropriately.

Reply:
updated